flang/lib/Lower/Bridge.cpp

```diff
@@ -216,13 +217,8 @@ class FirConverter : public Fortran::lower::AbstractConverter {
 public:
   explicit FirConverter(Fortran::lower::LoweringBridge &bridge,
                         fir::NameUniquer &uniquer)
-      : mlirContext{bridge.getMLIRContext()}, cooked{bridge.getCookedSource()},
-        module{bridge.getModule()}, defaults{bridge.getDefaultKinds()},
-        kindMap{bridge.getKindMap()}, uniquer{uniquer},
-        getShape{[&](const Fortran::lower::SomeExpr &expr) {
-          auto foldCtx = bridge.createFoldingContext();
-          return Fortran::evaluate::GetShape(foldCtx, expr);
-        }} {}
+      : bridge{bridge}, uniquer{uniquer}, foldingContext{
+                                              bridge.createFoldingContext()} {}
```

Comment:
I'm a little confused here. This is getting rid of all the data structures that were cached (including the partial application lambda) to call through to the builder. Then it goes and adds a cached copy of a FoldingContext.

Is it really valid to share a FoldingContext across different statements in the body of a function?

Reply:
FoldingContext context is actually rather expensive to create and this is my bad for making it this way. One of its member initializes a multimap containing lambdas to fold with the math runtime, so every time a FoldingContext is created, 300+ lambdas are created and placed inside a multimap (see https://github.com/flang-compiler/f18-llvm-project/blob/a7b319d7372a02a3f6f9019380eab73f91e46079/flang/lib/Evaluate/intrinsics-library.cpp#L395).  I could probably revisit this and see if there is a way to use the static_multimap (constexpr) instead. For now I feel like we should stay away from creating it every now and then.

I removed the other members because I added the bridge itself and I felt it made them a bit redundant. Do you prefer keeping them ? (I do not have a strong opinion here).

Reply:
My primary concern was that the FoldingContext might track "noise" from one context and propagate it erroneously to another context where it doesn't really apply. That concern stems from not fully understanding what FoldingContext is used for, etc. (The name implies it is caching information to me.)

Reply:
Also, have we measured the cost? I take it none of this cost can be turned to O(1) via `constexpr`. <sigh>

Reply:
I just measured the cost: the getShape FoldingContext creations can take **25% of the overall compilation time !** (granted, using dumb an example). Measurements detail below for information and records.
You are correct that the `FoldingContext` may accumulate junks: error and warning messages, so I agree it should be cheap to create and destroy. However, In the bridge we are not checking any of these front-end messages (These should not matter after semantic, when we do something in the bridge with a FoldingContext, we want to succeed.). Also, the front-end is mainly only using a single FoldingContext for everything (exception are spots where it want to attempt doing things while ignoring the error messages) so the existing pattern is not to create one for every use point.

Conclusion: I should fix `HostIntrinsicProceduresLibrary` to be either `constexpr` (they are a few challenges, but I think its doable) or make it live somewhere else so that it is not re-created (but I like this option less, its still a big state that has to be propagated everywhere). In the meantime I think we are safer having a single FoldingContext instance in the bridge.

Measurements:
Each call to `HostIntrinsicProceduresLibrary()` is on average (300 measures) ~ 0.5 ms, so 2000 creation of Folding context would cost 1s of compile time for something that only needs to be done once in 0.5 ms.
Assuming we are creating a context for every getShape, that means that 2000 array expression wast 1s compile time.
I did a stupid test with:
```
subroutine test(a)
  real :: a(10)
  a = a
  a = a
  ! repeat a = a 2000 times
end subroutine
```
bbc took 2333 ms (real time) to complete, and 836ms (cpu time) were spent in `HostIntrinsicProceduresLibrary()` called from creating FoldingContext (average of 10 runs), so that is 25% of the overall compilation time !
Given I am measuring on a debug version, the true proportion might even be worse...

Measurements were done with <chrono>

In evaluate/intrinsic-library.cpp
```
class OverallTimer {
public:
   OverallTimer(){}
   ~OverallTimer() {
      std::cerr << "spent total of: " << time.count() << " ms in " << nCall << " calls" << std::endl;
   }
   void start() {t1 = std::chrono::high_resolution_clock::now();}
   void stop() {
      t2 = std::chrono::high_resolution_clock::now();
      time += (t2-t1);
      nCall++;
   }
private:
   std::chrono::time_point<std::chrono::high_resolution_clock> t1, t2;
   std::chrono::duration<double, std::milli> time{decltype(time)::zero()};
   std::uint64_t nCall{0};
};
HostIntrinsicProceduresLibrary::HostIntrinsicProceduresLibrary() {
  static OverallTimer timer;
  timer.start();
  // code
 timer.stop();
}
```

Reply:
Wow. That is bad. Thanks for investigating this in more detail. That's very useful information.